### PR TITLE
feat(packageRules): matchQl

### PR DESCRIPTION
**Pull request by rarkins**
## Changes

Adds new `matchQl` capability for `packageRules`.

## Context

Closes #31662

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Also considering JSONata for this

File: docs/usage/configuration-options.md [hunk 1]
```diff
@@ -2969,6 +2969,34 @@ The above will set a replaceStrategy for any npm package which starts with `@ang
...
+It can be used to create matching on fields not supported by other `matchX` fields (e.g. `"matchQl": "groupName = 'foo'"`) or also as an alternative for existing fields.
+
+Principles:
+
+- strings must be surrounded in single `'` or double `"` quotes
+- numbers, boolean values, `null` and `undefined` should not use quotes
+- arrays can only contain strings
+- `string = string` (e.g. `manager = "npm"` or `packageName = "@foo/*"`) will use pattern matching (glob, regex, or exact string matching)
+- `string == string` (e.g. `manager == "npm"`) will use exact string matching only
+- `string = array[string]` (e.g. `manager = ["npm", "dockerfile"]`) will return true if _any_ of the array values pattern match
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2985]:
```suggestion
- `string = array[string]` (e.g. `manager = ["npm", "dockerfile"]`) returns True if _any_ of the array values pattern match
```

File: docs/usage/configuration-options.md [hunk 2]
```diff
@@ -2969,6 +2969,34 @@ The above will set a replaceStrategy for any npm package which starts with `@ang
...
+- `string != string` e.g. (`manager != "npm"`) returns true if it doesn't pattern match the string
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2986]:
```suggestion
- `string != string` e.g. (`manager != "npm"`) returns True if the string does _not_ match the pattern
```

File: docs/usage/configuration-options.md [hunk 3]
```diff
@@ -2969,6 +2969,34 @@ The above will set a replaceStrategy for any npm package which starts with `@ang
...
+- `string != array[string]` (e.g. `manager != ["npm", "dockerfile"]`) will return true if _none_ of the array values pattern match
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2987]:
```suggestion
- `string != array[string]` (e.g. `manager != ["npm", "dockerfile"]`) returns True if _none_ of the array values pattern match
```

File: docs/usage/configuration-options.md [hunk 4]
```diff
@@ -2969,6 +2969,34 @@ The above will set a replaceStrategy for any npm package which starts with `@ang
...
+- `number` values (e.g. `currentVersionAge`, which is measured in days) support comparisons `=`, `!=`, `<`, `<=`, `>` and `>=`
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2988]:
```suggestion
- `number` values support comparisons `=`, `!=`, `<`, `<=`, `>` and `>=`, this comparison can be helpful for fields like `currentVersionAge` (which is measured in days)
```

Let's put the comparison first, and _then_ show the example. And expand on the example a bit.

File: docs/usage/configuration-options.md [hunk 5]
```diff
@@ -2969,6 +2969,34 @@ The above will set a replaceStrategy for any npm package which starts with `@ang
...
+- `boolean` values support `=` and `!=` comparisons
+
+Here are some equivalent statements:
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2991]:
```suggestion
#### Equivalent statements
```

This should have it's own sub-heading too.

File: docs/usage/configuration-options.md [hunk 6]
```diff
@@ -2969,6 +2969,34 @@ The above will set a replaceStrategy for any npm package which starts with `@ang
...
+
+| Original                            | Match equivalent                          |
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:2993]:
```suggestion
| Original rule                       | `MatchQl` equivalent                      |
```

File: lib/config/options/index.ts [hunk 7]
```diff
@@ -1248,6 +1248,16 @@ const options: RenovateOptions[] = [
     stage: 'package',
     mergeable: true,
   },
+  {
+    name: 'matchQl',
+    description: 'SQL-like query language for matching',
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:1253]:
```suggestion
    description: 'SQL-like query language for matching.',
```

File: lib/config/options/index.ts [hunk 8]
```diff
@@ -1248,6 +1248,16 @@ const options: RenovateOptions[] = [
...
+    type: 'string',
+    parents: ['packageRules'],
+    stage: 'package',
+    mergeable: true,
+    cli: false,
+    env: false,
```
Comment (HonkingGoose) [inline lib/config/options/index.ts:1259]: The docs say the feature is _experimental_, but there's no `experimental: true` field in the code.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
Replaced by https://github.com/renovatebot/renovate/pull/31826

**Pull request closed without merging**